lib/codecs/src/decoding/framing/chunked_gelf.rs

```diff
@@ -0,0 +1,881 @@
+use crate::{BytesDecoder, StreamDecodingError};
+
+use super::{BoxedFramingError, FramingError};
+use bytes::{Buf, Bytes, BytesMut};
+use derivative::Derivative;
+use snafu::{ensure, Snafu};
+use std::any::Any;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+use tokio;
+use tokio::task::JoinHandle;
+use tokio_util::codec::Decoder;
+use tracing::{debug, warn};
+use vector_config::configurable_component;
+
+const GELF_MAGIC: &[u8] = &[0x1e, 0x0f];
+const GELF_MAX_TOTAL_CHUNKS: u8 = 128;
+const DEFAULT_TIMEOUT_SECS: f64 = 5.0;
+
+const fn default_timeout_secs() -> f64 {
+    DEFAULT_TIMEOUT_SECS
+}
+
+/// Config used to build a `ChunkedGelfDecoder`.
+#[configurable_component]
+#[derive(Debug, Clone, Default)]
+pub struct ChunkedGelfDecoderConfig {
+    /// Options for the chunked GELF decoder.
+    #[serde(default)]
+    pub chunked_gelf: ChunkedGelfDecoderOptions,
+}
+
+impl ChunkedGelfDecoderConfig {
+    /// Build the `ChunkedGelfDecoder` from this configuration.
+    pub fn build(&self) -> ChunkedGelfDecoder {
+        ChunkedGelfDecoder::new(
+            self.chunked_gelf.timeout_secs,
+            self.chunked_gelf.pending_messages_limit,
+            self.chunked_gelf.max_chunk_length,
+            self.chunked_gelf.max_message_length,
+        )
+    }
+}
+
+/// Options for building a `ChunkedGelfDecoder`.
+#[configurable_component]
+#[derive(Clone, Debug, Derivative)]
+#[derivative(Default)]
+pub struct ChunkedGelfDecoderOptions {
+    /// The timeout, in seconds, for a message to be fully received. If the timeout is reached, the
+    /// decoder drops all the received chunks of the timed out message.
+    #[serde(default = "default_timeout_secs")]
+    #[derivative(Default(value = "default_timeout_secs()"))]
+    pub timeout_secs: f64,
+
+    /// The maximum number of pending incomplete messages. If this limit is reached, the decoder starts
+    /// dropping chunks of new messages, ensuring the memory usage of the decoder's state is bounded.
+    /// If this option is not set, the decoder does not limit the number of pending messages and the memory usage
+    /// of its messages buffer can grow unbounded. This matches Graylog Server's behavior.
+    #[serde(default, skip_serializing_if = "vector_core::serde::is_default")]
+    pub pending_messages_limit: Option<usize>,
+
+    /// The maximum length of a single GELF chunk, in bytes. Chunks longer than this length will
+    /// be dropped. If this option is not set, the decoder does not limit the length of chunks and
+    /// the per-chunk memory is unbounded.
+    ///
+    /// This limit takes only into account the chunk's payload and the GELF header bytes are excluded from the calculation.
+    #[serde(skip_serializing_if = "vector_core::serde::is_default")]
+    pub max_chunk_length: Option<usize>,
```

Comment:
Should we use no limit by default (`None`) Or use the max packet size of UDP datagrams of `65,535` bytes? (actually, we should use `65,535 - 10`, because for this chunk length count we exclude the 10 bytes of the chunked gelf header

Reply:
Given the framers have a convention of defaulting to no max length, I think we can leave the default for this as `None` too. I also expect people will more often want to limit the total length of the message, rather than each chunk.